File: backup/37,5million/password_cracker.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// Wandelt einen Index in ein Passwort der Länge 'length' um.
func indexToPassword(idx int, charset []rune, length int) string {
	password := make([]rune, length)
	csLen := len(charset)
	for i := length - 1; i >= 0; i-- {
		password[i] = charset[idx%csLen]
		idx /= csLen
	}
	return string(password)
}

// Berechnet die Gesamtzahl der Kombinationen für eine gegebene Länge.
func totalForLength(length, csLen int) int64 {
	total := int64(1)
	for i := 0; i < length; i++ {
		total *= int64(csLen)
	}
	return total
}

// Struktur für einen Task, der einen bestimmten Indexbereich abdeckt.
type task struct {
	length int
	start  int
	end    int
}

// Worker, der Tasks aus der Channel-Pipeline verarbeitet.
func worker(tasks <-chan task, charset []rune, target string, foundFlag *int32, testedCounter *int64, wg *sync.WaitGroup) {
	defer wg.Done()
	batchSize := 10000
	for t := range tasks {
		// Falls das Passwort bereits gefunden wurde, Task überspringen.
		if atomic.LoadInt32(foundFlag) == 1 {
			continue
		}
		for idx := t.start; idx <= t.end; idx += batchSize {
			if atomic.LoadInt32(foundFlag) == 1 {
				break
			}
			endBatch := t.end
			if idx+batchSize-1 < t.end {
				endBatch = idx + batchSize - 1
			}
			atomic.AddInt64(testedCounter, int64(endBatch-idx+1))
			for i := idx; i <= endBatch; i++ {
				if indexToPassword(i, charset, t.length) == target {
					atomic.StoreInt32(foundFlag, 1)
					return
				}
			}
		}
	}
}

// Formatiert eine Zahl mit Punkten als Tausendertrennzeichen
func formatNumber(number int64) string {
	in := strconv.FormatInt(number, 10)
	out := ""
	for i, digit := range in {
		if i > 0 && (len(in)-i)%3 == 0 {
			out += "."
		}
		out += string(digit)
	}
	return out
}

// Berechnet die geschätzte Zeit basierend auf der Anzahl der Kombinationen und der Geschwindigkeit
func calculateEstimatedTime(totalCombinations, speed int64) (seconds, hours, days, years float64) {
	seconds = float64(totalCombinations) / float64(speed)
	hours = seconds / 3600
	days = hours / 24
	years = days / 365
	return
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Zielpasswort: ")
	target, _ := reader.ReadString('\n')
	target = strings.TrimSpace(target)

	fmt.Println("\nZeichensatzauswahl:")
	fmt.Println("1. Nur Zahlen")
	fmt.Println("2. Nur Buchstaben")
	fmt.Println("3. Buchstaben + Zahlen")
	fmt.Println("4. Alle Zeichen")
	fmt.Print("Wahl (1-4): ")
	choice, _ := reader.ReadString('\n')
	choice = strings.TrimSpace(choice)

	var charset []rune
	switch choice {
	case "1":
		charset = []rune("0123456789")
	case "2":
		charset = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
	case "3":
		charset = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
	case "4":
		charset = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")
	default:
		fmt.Println("Ungültige Eingabe!")
		return
	}

	fmt.Print("Maximale Länge: ")
	maxLengthStr, _ := reader.ReadString('\n')
	maxLengthStr = strings.TrimSpace(maxLengthStr)
	maxLength, err := strconv.Atoi(maxLengthStr)
	if err != nil || maxLength <= 0 {
		fmt.Println("Ungültige Länge!")
		return
	}

	// Berechne die Gesamtzahl der möglichen Kombinationen für die maximale Länge
	csLen := len(charset)
	totalCombinations := totalForLength(maxLength, csLen)
	fmt.Printf("Gesamtzahl der möglichen Kombinationen für eine Länge von %d: %s\n", maxLength, formatNumber(totalCombinations))

	// Beispielgeschwindigkeit: 37 Millionen Passwörter pro Sekunde
	speed := int64(28000000)
	seconds, hours, days, years := calculateEstimatedTime(totalCombinations, speed)
	fmt.Printf("Geschätzte Zeit für das Durchprobieren aller Kombinationen bei %d Passwörtern/Sekunde:\n", speed)
	fmt.Printf("Sekunden: %.2f\n", seconds)
	fmt.Printf("Stunden: %.2f\n", hours)
	fmt.Printf("Tage: %.2f\n", days)
	fmt.Printf("Jahre: %.2f\n", years)

	startTime := time.Now()
	var foundFlag int32 = 0
	var testedCounter int64 = 0

	numCPU := runtime.NumCPU()
	runtime.GOMAXPROCS(numCPU)

	// Task-Channel und Worker-Pool starten.
	tasks := make(chan task, numCPU*2)
	var wg sync.WaitGroup
	for i := 0; i < numCPU; i++ {
		wg.Add(1)
		go worker(tasks, charset, target, &foundFlag, &testedCounter, &wg)
	}

	// Fortschrittsanzeige in einem separaten Goroutine.
	done := make(chan struct{})
	go func() {
		for {
			select {
			case <-done:
				return
			default:
				fmt.Printf("\rGesamtfortschritt: %s Passwörter getestet", formatNumber(atomic.LoadInt64(&testedCounter)))
				time.Sleep(500 * time.Millisecond)
			}
		}
	}()

	// Erzeuge Tasks für jede Passwortlänge.
	for length := 1; length <= maxLength && atomic.LoadInt32(&foundFlag) == 0; length++ {
		total := totalForLength(length, csLen)
		// Bestimme die Chunk-Größe: maximal 10000 oder total/(10*numCPU), je nachdem, was größer ist.
		chunkSize := 10000
		temp := int(total) / (10 * numCPU)
		if temp > chunkSize {
			chunkSize = temp
		}
		for start := 0; start < int(total) && atomic.LoadInt32(&foundFlag) == 0; start += chunkSize {
			end := start + chunkSize - 1
			if end >= int(total) {
				end = int(total) - 1
			}
			tasks <- task{length: length, start: start, end: end}
		}
	}
	close(tasks)
	wg.Wait()
	close(done)

	elapsedTime := time.Since(startTime).Seconds()
	fmt.Println("\n" + strings.Repeat("=", 50))
	if atomic.LoadInt32(&foundFlag) == 1 {
		fmt.Printf("Passwort gefunden!\n")
		fmt.Printf("Gesamtzeit: %.2fs\n", elapsedTime)
		fmt.Printf("Getestete Passwörter: %s\n", formatNumber(atomic.LoadInt64(&testedCounter)))
		fmt.Printf("Durchschnittliche Geschwindigkeit: %s Passwörter/Sekunde\n", formatNumber(int64(float64(atomic.LoadInt64(&testedCounter))/elapsedTime)))
	} else {
		fmt.Println("Passwort nicht gefunden")
	}
}

File: backup/300million/password_cracker.py
import string
import time
import math
from multiprocessing import Pool, cpu_count, Manager
from tqdm import tqdm

def index_to_password(idx, charset, length):
    password = []
    cs_len = len(charset)
    for _ in range(length):
        idx, rem = divmod(idx, cs_len)
        password.append(charset[rem])
    return ''.join(reversed(password))

def check_range(args):
    start, end, charset, length, target, found_flag = args
    if found_flag.value:
        return 0, 0
    
    cs_len = len(charset)
    batch_size = 5000
    tested = 0
    
    for idx in range(start, end + 1, batch_size):
        if found_flag.value:
            return tested, 0
        
        batch_end = min(idx + batch_size - 1, end)
        tested += batch_end - idx + 1
        
        for i in range(idx, batch_end + 1):
            if index_to_password(i, charset, length) == target:
                found_flag.value = 1
                return tested, 1
    
    return tested, 0

def main():
    target = input("Zielpasswort: ")
    
    print("\nZeichensatzauswahl:")
    print("1. Nur Zahlen\n2. Nur Buchstaben\n3. Buchstaben + Zahlen\n4. Alle Zeichen")
    charset = {
        '1': string.digits,
        '2': string.ascii_letters,
        '3': string.ascii_letters + string.digits,
        '4': string.ascii_letters + string.digits + string.punctuation
    }.get(input("Wahl (1-4): "), "")
    
    if not charset:
        print("Ungültige Eingabe!")
        return
    
    max_length = int(input("Maximale Länge: "))
    charset = tuple(charset)
    cs_len = len(charset)
    total_combinations = sum(cs_len**l for l in range(1, max_length+1))
    
    start_time = time.time()
    manager = Manager()
    found_flag = manager.Value('i', 0)
    total_tested = 0
    
    with tqdm(total=total_combinations, desc="Gesamtfortschritt", unit="pwd",
             bar_format="{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}{postfix}]") as pbar:
        with Pool(processes=cpu_count()) as pool:
            for length in range(1, max_length + 1):
                if found_flag.value:
                    break
                
                total = cs_len ** length
                chunk_size = max(1000, total // (cpu_count() * 10))
                chunks = []
                
                for start in range(0, total, chunk_size):
                    end = min(start + chunk_size - 1, total - 1)
                    chunks.append((
                        start, 
                        end, 
                        charset, 
                        length, 
                        target, 
                        found_flag
                    ))
                
                results = pool.imap_unordered(check_range, chunks, chunksize=10)
                
                for tested, found in results:
                    total_tested += tested
                    pbar.update(tested)
                    if found:
                        found_flag.value = 1
                        pool.terminate()
                        break

    print("\n" + "="*50)
    print(f"Passwort {'gefunden' if found_flag.value else 'nicht gefunden'}")
    
    if found_flag.value:
        elapsed_time = time.time() - start_time
        print(f"Gesamtzeit: {elapsed_time:.2f}s")
        print(f"Getestete Passwörter: {total_tested}")
        print(f"Durchschnittliche Geschwindigkeit: {total_tested / elapsed_time:,.0f} Passwörter/Sekunde")

if __name__ == "__main__":
    main()